bus-mapping/src/circuit_input_builder.rs

```diff
@@ -181,101 +217,136 @@ pub enum CallType {
     Create2,
 }
 
-impl CallType {
+impl CallKind {
     fn is_create(&self) -> bool {
         matches!(self, Self::Create | Self::Create2)
     }
 }
 
-impl TryFrom<OpcodeId> for CallType {
+impl TryFrom<OpcodeId> for CallKind {
     type Error = Error;
 
     fn try_from(op: OpcodeId) -> Result<Self, Self::Error> {
         Ok(match op {
-            OpcodeId::CALL => CallType::Call,
-            OpcodeId::CALLCODE => CallType::CallCode,
-            OpcodeId::DELEGATECALL => CallType::DelegateCall,
-            OpcodeId::STATICCALL => CallType::StaticCall,
-            OpcodeId::CREATE => CallType::Create,
-            OpcodeId::CREATE2 => CallType::Create2,
+            OpcodeId::CALL => CallKind::Call,
+            OpcodeId::CALLCODE => CallKind::CallCode,
+            OpcodeId::DELEGATECALL => CallKind::DelegateCall,
+            OpcodeId::STATICCALL => CallKind::StaticCall,
+            OpcodeId::CREATE => CallKind::Create,
+            OpcodeId::CREATE2 => CallKind::Create2,
             _ => return Err(Error::OpcodeIdNotCallType),
         })
     }
 }
 
-/// Context of a Call during a [`Transaction`] which can mutate in an
-/// [`ExecStep`].
+/// Circuit Input related to an Ethereum Call
 #[derive(Debug)]
-pub struct CallContext {
+pub struct Call {
     /// Type of call
-    call: CallType,
+    kind: CallKind,
     /// This call is being executed without write access (STATIC)
     is_static: bool,
     /// This call is root call with tx.to == null, or op == CREATE or op ==
     /// CREATE2
     is_create: bool,
+    /// This call generated implicity by a Transaction.
+    is_root: bool,
     /// Address where this call is being executed
     pub address: Address,
+    /// Code Hash
+    code_hash: H256,
 }
 
 #[derive(Debug)]
 /// Context of a [`Transaction`] which can mutate in an [`ExecStep`].
 pub struct TransactionContext {
-    call_stack: Vec<CallContext>,
+    /// Call Stack by indices.
+    /// The call_stack will always have a fixed elemnt at index 0 which
+    /// corresponds to the call implicitly created by the transaction.
+    call_stack: Vec<usize>,
+    /// State Write Counter
+    swc: usize,
```

Comment:
I think `swc`'s lifecycle is inside a reversion section (a reverted call skipping reverted sub-calls) instead of a tx.

Reply:
I agree. It should be per call that hasn't yet exited with `RETURN`.

Reply:
To clarify after a call with @ed255. This should live inside of the `CallContext` struct. And we should have it for each call of the trace.

Reply:
I've added the swc to a `CallContext`, and not removed it from `ExecStep` yet because it actually appears in the circuit `ExecStep`: https://github.com/han0110/zkevm-circuits/blob/32fbc01235e36e7af665df89e48d6c2217a8d4e7/zkevm-circuits/src/evm_circuit/execution.rs#L195
Let's discuss this tomorrow in a call!

Reply:
@han0110 I've updated the PR following the idea we discussed yesterday.  In every call, the swc starts at 0, and after a return, if the subcall suceeds the swc gets incremented by the swc of successful writes from the subcall.

I've also added some code in the circuit input builder when handling each step so that *CALL* and return from a function is identified to apply this logic.  Let me know what you think!